README.md
A wrapper for raspistill and raspivid (todo).

Not all the cli options are implemented please PR options that you need.

On ubuntu to may need to loosen the permissions of /dev/vchiq.
````sh
$ chmod 666 /dev/vchiq
````

CHANGELOG.md
## 1.0.0

- Initial version.

## 1.0.1

- Add library tag.
- Export options as well as raspicam.


test/runner.dart
import 'dart:convert';
import 'dart:io';

import 'package:raspicam/raspicam.dart';
import 'package:raspicam/raspicam_options.dart';

void main() {
  runner();
}

Stream<String> readLine() =>
    stdin.transform(utf8.decoder).transform(const LineSplitter());

void processLine(String line) {
  print(line);
}

void runner() async {
  var opts = RaspicamOptions();
  // opts.width = 2464;
  // opts.height = 2464;
  opts.width = 360;
  opts.height = 360;
  opts.previewHeight = 640;
  opts.previewX = 1200;
  opts.awb = 'off';
  opts.datetime = true;
  opts.folder = './test';

  print(opts());

  var rpc = Raspicam(opts);
  await rpc.start();
  var p1 = await rpc.takePhoto();
  print('Photo 1 ${p1}');

  readLine().listen((line) async {
    if (line.contains('p')) {
      var myPhoto = await rpc.takePhoto();
      print(myPhoto);
    } else {
      rpc.killChildProcess();
    }
  });
}


test/raspicam_test.dart
// @dart=2.9
import 'package:raspicam/src/raspicam_options.dart';
import 'package:test/test.dart';
import 'package:raspicam/src/raspicam.dart';

void main() {
  test('raspicam', () async {
    var opts = RaspicamOptions();
    var rpc = Raspicam(opts);
    await rpc.start();
    var myImage = await rpc.takePhoto();
    rpc.stop();
    expect(myImage.contains('image'), true);
  });
}


lib/src/raspicam.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:raspicam/src/raspicam_options.dart';

class Raspicam {
  String executable = 'raspistill';
  bool isRunning = false;
  bool isStarted = false;
  bool takingPhoto = false;
  late Process childProcess;
  RaspicamOptions options;

  String _currentPhoto = '';
  late DateTime _endTime;
  late DateTime _startTime;
  Completer _readyCompleter = Completer();
  Completer _takePhotoCompleter = Completer();

  Raspicam(this.options);

  Future<dynamic> start() async {
    await destroyPreviousPiCameraProcesses();
    childProcess = await Process.start(executable, options());

    isStarted = true;
    attachListeners();
    _takePhotoCompleter.complete('');
    return _readyCompleter.future;
  }

  void stop() {
    childProcess.kill();
  }

  Future<dynamic> takePhoto() {
    if (!isRunning) {
      throw Exception('Camera is not running.');
    }

    if (!_takePhotoCompleter.isCompleted) {
      throw Exception('Still processing last photo.');
    }

    _takePhotoCompleter = Completer();

    childProcess.kill(ProcessSignal.sigusr1);

    return _takePhotoCompleter.future;
  }

  Future<ProcessResult> destroyPreviousPiCameraProcesses() async {
    return Process.run('pkill', ['raspistill']);
  }

  void attachListeners() {
    childProcess.stderr.transform(utf8.decoder).listen((data) {
      print(data);
      isRunning = isRunning || checkReady(data);

      if (isRunning && !_readyCompleter.isCompleted) {
        _readyCompleter.complete(childProcess);
      }

      if (openingFile(data)) {
        _currentPhoto = parseOpeningFile(data);
      }

      if (checkStartingCapture(data)) {
        _startTime = DateTime.now();
      }

      if (checkFinishedCapture(data)) {
        _endTime = DateTime.now();
        print(_endTime.difference(_startTime));
        _takePhotoCompleter.complete(_currentPhoto);
      }
    });

    childProcess.exitCode.then((code) {
      print('Got an exit code $code');
      isRunning = false;
      isStarted = false;
      if (!_takePhotoCompleter.isCompleted) {
        _takePhotoCompleter
            .completeError(Exception('Camera was killed while taking photo.'));
      }
    });
  }

  bool checkReady(String data) {
    return data.contains('Waiting for SIGUSR');
  }

  bool checkStartingCapture(String data) {
    return data.contains('Starting capture');
  }

  bool checkFinishedCapture(String data) {
    return data.contains('Finished capture');
  }

  bool openingFile(String data) {
    return data.contains('Opening output file');
  }

  String parseOpeningFile(String data) {
    return data.replaceAll('Opening output file', '').trim();
  }
}


lib/src/raspicam_options.dart
import 'package:path/path.dart' as path;

/// Not all options are implemented. Please PR any options you need added.
class RaspicamOptions {
  int quality = 98;
  String encoding = 'jpg';
  bool datetime = false;
  bool timestamp = false;

  int width = 640;
  int height = 480;
  String output = '';
  String folder = '';
  String prefix = 'image';
  int leadingZeros = 4;

  int previewX = 0;
  int previewY = 0;
  int previewWidth = 640;
  int previewHeight = 480;
  bool noPreview = false;

  int ISO = 0;
  String exposure = 'auto';
  String awb = 'auto';
  bool hflip = true;
  bool vflip = true;
  double framesPerSecond = 30;
  double shutter = 0; //1 / 30 * 1e6;

  // Gains are ignored if exposure mode and/or ISO is set
  List<double> awbGains = [744 / 256, 489 / 256];
  double analogGain = 640 / 256;
  double digitalGain = 320 / 256;
  bool settings = false;

  List<String> call() {
    return optionsToList();
  }

  List<String> optionsToList() {
    if (shutter == 0 && framesPerSecond == 0) framesPerSecond = 30;

    return [
      '--verbose',
      '--signal',
      '--quality',
      quality.toString(),
      '--encoding',
      encoding,
      if (datetime) '--datetime',
      if (!datetime && timestamp) '--timestamp',
      '--width',
      width.toString(),
      '--height',
      height.toString(),
      '--output',
      output.isNotEmpty
          ? path.join(folder, output)
          : path.join(folder, '${prefix}_%0${leadingZeros}d.${encoding}'),
      if (noPreview) '--nopreview',
      if (!noPreview) '--preview',
      if (!noPreview)
        '${previewX},${previewY},${previewWidth},${previewHeight}',
      if (ISO > 0) '--ISO',
      if (ISO > 0) ISO.toString(),
      if (exposure != 'auto') '--exposure',
      if (exposure != 'auto') exposure,
      '--awb',
      awb,
      if (hflip) '--hflip',
      if (vflip) '--vflip',
      '--shutter',
      shutter > 0 ? shutter.toString() : (1 / framesPerSecond * 1e6).toString(),
      if (awb == 'off') '--awbgains',
      if (awb == 'off') awbGains.join(','),
      if (analogGain != 0) '--analoggain',
      if (analogGain != 0) analogGain.toString(),
      if (digitalGain != 0) '--digitalgain',
      if (digitalGain != 0) digitalGain.toString(),
    ];
  }
}


example/example.dart
import 'dart:convert';
import 'dart:io';

import 'package:raspicam/src/raspicam.dart';
import 'package:raspicam/src/raspicam_options.dart';

void main() {
  runner();
}

Stream<String> readLine() =>
    stdin.transform(utf8.decoder).transform(const LineSplitter());

void processLine(String line) {
  print(line);
}

void runner() async {
  var opts = RaspicamOptions();
  opts.width = 360;
  opts.height = 360;
  opts.previewHeight = 640;
  opts.previewX = 1200;
  opts.awb = 'off';
  opts.datetime = true;
  opts.folder = './test';

  print(opts());

  var rpc = Raspicam(opts);
  await rpc.start();
  var p1 = await rpc.takePhoto();
  print('Took a photo ${p1}');

  print('Press p to take more photos, anything else to exit.');
  readLine().listen((line) async {
    if (line.contains('p')) {
      var myPhoto = await rpc.takePhoto();
      print(myPhoto);
    } else {
      rpc.stop();
    }
  });
}
